components/movementsensor/imuvectornav/imu.go

```diff
@@ -122,33 +117,66 @@ func newVectorNav(
 		return nil, err
 	}
 
-	speed := *newConf.Speed
-	if speed == 0 {
-		speed = 8000000
-	}
+	baudRate, pollFreq, gpioPin := configParsingAndSetup(ctx, logger, newConf)
 
-	pfreq := *newConf.Pfreq
 	v := &vectornav{
 		Named:     conf.ResourceName().AsNamed(),
 		bus:       buses.NewSpiBus(newConf.SPI),
 		logger:    logger,
-		cs:        newConf.CSPin,
-		speed:     speed,
+		gpioPin:   gpioPin,
+		baudRate:  baudRate,
 		busClosed: false,
-		polling:   pfreq,
+		polling:   uint(pollFreq),
 	}
-	mdl, err := v.readRegisterSPI(ctx, modelNumber, 24)
+
+	err = v.registerReading(ctx, logger)
 	if err != nil {
 		return nil, err
 	}
-	sn, err := v.readRegisterSPI(ctx, serialNumber, 4)
+
+	err = v.imuConfig(ctx)
 	if err != nil {
 		return nil, err
 	}
-	fwver, err := v.readRegisterSPI(ctx, firmwareVersion, 4)
```

Comment:
I actually prefer the flatter version of the constructor; I do not like setting things from within the helper functions because of potential locks with a spi communication protocol, which the Linux kernel fdopen does not ensure thread safety with. So please make it flatter again.